components/style/element_flags.rs

```diff
@@ -20,5 +22,43 @@ bitflags! {
         /// last children must be restyled, because they may match :first-child,
         /// :last-child, or :only-child.
         const HAS_EDGE_CHILD_SELECTOR = 1 << 2,
+        /// When this element is affected by the empty selector.
+        const HAS_EMPTY_SELECTOR = 1 << 3,
+    }
+}
+
+impl ElementFlags {
+    /// Returns the two sets of flags (one for the child, one for the parent)
+    /// generated for this set of relations.
+    #[inline]
+    pub fn from_relations(relations: StyleRelations) -> (Self, Self) {
+        use selectors::matching::*;
+        let mut child = Self::empty();
+        let mut parent = Self::empty();
+        if relations.intersects(AFFECTED_BY_EMPTY) {
+            // FIXME(emilio): This doesn't need to be atomic.
```

Comment:
This comment is duplicated below, which is probably a more appropriate place for it.